File: src/deadlock_build_sync/quality_replay.py
"""Closed, deidentified inputs for later evaluation of a frozen policy."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from .recommendation_state import (
    DecisionState,
    RecommendationAction,
    RecommendationError,
)
from .value_validation import object_dict, object_rows

if TYPE_CHECKING:
    from .policy import BuildPolicy


@dataclass(frozen=True)
class ReplayCase:
    policy_id: str
    match_group: str
    state: DecisionState
    observed_action: RecommendationAction
    observed_item_id: int | None
    core_completed: bool
    behind: bool
    ambiguous_purchase: bool


def _require_positive_integer(value: object, label: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value < 1:
        raise RecommendationError(f"replay has invalid {label}")
    return value


def _require_boolean_field(row: dict[str, object], key: str) -> bool:
    value = row.get(key)
    if not isinstance(value, bool):
        raise RecommendationError(f"replay requires boolean {key}")
    return value


def _parse_replay_case(
    row: dict[str, object], policy: BuildPolicy, cutoff: int
) -> ReplayCase:
    state = DecisionState.from_document(row.get("state"))
    if state.hero_id != policy.hero_id:
        raise RecommendationError("replay hero differs from frozen policy")
    start = _require_positive_integer(row.get("match_start_timestamp"), "match start")
    observed_at = _require_positive_integer(
        row.get("feature_as_of_timestamp"), "feature timestamp"
    )
    assigned_at = _require_positive_integer(
        row.get("policy_assigned_at"), "policy assignment"
    )
    if start <= cutoff:
        raise RecommendationError(
            "independent replay must start after the frozen cutoff"
        )
    if not start <= observed_at <= start + state.clock_s:
        raise RecommendationError("replay features are outside the pre-decision window")
    if not cutoff < assigned_at <= start:
        raise RecommendationError("replay policy must be assigned before the match")
    group = row.get("match_group")
    if not isinstance(group, str) or not group.strip():
        raise RecommendationError("replay lacks a deidentified match group")
    try:
        action = RecommendationAction(str(row.get("observed_action")))
    except ValueError as error:
        raise RecommendationError("replay has invalid observed action") from error
    if "observed_item_id" not in row:
        raise RecommendationError("replay requires observed_item_id, null for non-buys")
    item = row.get("observed_item_id")
    if action == RecommendationAction.BUY:
        item = _require_positive_integer(item, "observed item")
    elif item is not None:
        raise RecommendationError("only observed buys may name an item")
    return ReplayCase(
        policy.policy_id,
        group,
        state,
        action,
        item,
        _require_boolean_field(row, "core_completed"),
        _require_boolean_field(row, "behind"),
        _require_boolean_field(row, "ambiguous_purchase"),
    )


def parse_replay(
    document: object,
    policies: tuple[BuildPolicy, ...],
    *,
    cutoff: int,
    evidence_id: str,
) -> tuple[ReplayCase, ...]:
    """Reject reused, future-derived, duplicate, or cross-policy replay inputs.

    Returns:
        Cases retaining unfinished builds and ambiguous purchases for coverage audits.

    Raises:
        RecommendationError: If the replay violates its closed contract.

    """
    data = object_dict(document)
    if (
        data is None
        or type(data.get("schema_version")) is not int
        or data.get("schema_version") != 1
    ):
        raise RecommendationError("unsupported quality replay schema")
    if data.get("cohort_selection") != "all_eligible_player_matches":
        raise RecommendationError("replay must retain unfinished builds")
    raw_rows = data.get("cases")
    rows = object_rows(raw_rows)
    if rows is None or not isinstance(raw_rows, list) or len(rows) != len(raw_rows):
        raise RecommendationError("replay cases must be a list of objects")
    by_id = {policy.policy_id: policy for policy in policies}
    cases = []
    seen: set[tuple[str, str, int, int]] = set()
    for row in rows:
        policy = by_id.get(str(row.get("policy_id")))
        if policy is None:
            raise RecommendationError("replay references another frozen policy")
        case = _parse_replay_case(row, policy, cutoff)
        if case.state.build_evidence_id != evidence_id:
            raise RecommendationError("replay references another evidence artifact")
        key = (case.policy_id, case.match_group, case.state.hero_id, case.state.clock_s)
        if key in seen:
            raise RecommendationError("replay repeats a decision opportunity")
        seen.add(key)
        cases.append(case)
    return tuple(cases)

File: crates/deadlock-guides/src/quality_replay.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::{Error, Result, array, integer, text};
use serde_json::Value;

use crate::decision_state::DecisionState;
use crate::policy_artifact::PolicyArtifact;
use crate::policy_model::BuildPolicy;
use crate::recommendation_types::RecommendationAction;

#[derive(Debug)]
pub struct ReplayCase {
    pub policy_id: String,
    pub match_group: String,
    pub state: DecisionState,
    pub observed_action: RecommendationAction,
    pub observed_item_id: Option<u64>,
    pub core_completed: bool,
    pub behind: bool,
    pub ambiguous_purchase: bool,
}

/// # Errors
/// Returns an error when replay reuses training observations or contains incompatible, malformed, repeated, or future-derived decisions.
pub fn parse_replay(
    document: &Value,
    policies: &PolicyArtifact,
    cutoff: i64,
    evidence_id: &str,
) -> Result<Vec<ReplayCase>> {
    if document["schema_version"].as_u64() != Some(1)
        || document["cohort_selection"] != "all_eligible_player_matches"
    {
        return Err(Error::new(
            "Replay requires schema 1 and all eligible player matches",
        ));
    }
    let by_id = policies
        .policies()
        .values()
        .map(|policy| (policy.policy_id(), policy))
        .collect::<BTreeMap<_, _>>();
    let mut seen = BTreeSet::new();
    let mut cases = Vec::new();
    for row in array(&document["cases"])? {
        let id = text(row, "policy_id")?;
        let policy = by_id
            .get(id)
            .ok_or_else(|| Error::new("Replay references another frozen policy"))?;
        let case = parse_case(row, policy, cutoff)?;
        if case.state.content().build_evidence_id != evidence_id {
            return Err(Error::new("Replay references another evidence artifact"));
        }
        if !seen.insert((
            case.policy_id.clone(),
            case.match_group.clone(),
            case.state.content().hero_id,
            case.state.content().clock_s,
        )) {
            return Err(Error::new("Replay repeats a decision opportunity"));
        }
        cases.push(case);
    }
    Ok(cases)
}

fn parse_case(row: &Value, policy: &BuildPolicy, cutoff: i64) -> Result<ReplayCase> {
    let state = DecisionState::from_document(&row["state"])?;
    if state.content().hero_id != policy.content().hero_id {
        return Err(Error::new("Replay hero differs from the frozen policy"));
    }
    validate_timestamps(row, &state, cutoff)?;
    let group = text(row, "match_group")?;
    if group.trim().is_empty() {
        return Err(Error::new("Replay lacks a deidentified match group"));
    }
    let action: RecommendationAction = serde_json::from_value(row["observed_action"].clone())?;
    let item = row.get("observed_item_id").ok_or_else(|| {
        Error::new("Replay requires observed_item_id, with null for non-purchase actions")
    })?;
    let item =
        if action == RecommendationAction::Buy {
            Some(item.as_u64().filter(|id| *id > 0).ok_or_else(|| {
                Error::new("Observed purchase requires a positive item identifier")
            })?)
        } else if item.is_null() {
            None
        } else {
            return Err(Error::new("Only observed purchases can name an item"));
        };
    Ok(ReplayCase {
        policy_id: policy.policy_id().into(),
        match_group: group.into(),
        state,
        observed_action: action,
        observed_item_id: item,
        core_completed: boolean(row, "core_completed")?,
        behind: boolean(row, "behind")?,
        ambiguous_purchase: boolean(row, "ambiguous_purchase")?,
    })
}

fn validate_timestamps(row: &Value, state: &DecisionState, cutoff: i64) -> Result<()> {
    let start = integer(row, "match_start_timestamp")?;
    let observed = integer(row, "feature_as_of_timestamp")?;
    let assigned = integer(row, "policy_assigned_at")?;
    let cutoff = u64::try_from(cutoff)?;
    let decision_time = start
        .checked_add(state.content().clock_s)
        .ok_or_else(|| Error::new("Replay decision timestamp exceeds 64 bits"))?;
    if start <= cutoff
        || assigned <= cutoff
        || assigned > start
        || !(start..=decision_time).contains(&observed)
    {
        return Err(Error::new(
            "Replay timestamps violate the frozen cutoff, policy assignment, or decision window",
        ));
    }
    Ok(())
}

fn boolean(row: &Value, name: &str) -> Result<bool> {
    row[name]
        .as_bool()
        .ok_or_else(|| Error::new(format!("Replay requires boolean {name}")))
}
